samples/ClassroomLabs/Modules/Library/Tools/LabCreationLibrary.psm1

```diff
@@ -0,0 +1,598 @@
+    [CmdletBinding()]
+    param(
+        [parameter(Mandatory = $true, HelpMessage = "Array containing one line for each lab to be created", ValueFromPipeline = $true)]
+        [ValidateNotNullOrEmpty()]
+        [psobject[]]
+        $labs,
+
+        [Parameter(Mandatory = $false, HelpMessage = "Pick one lab from the labs' list")]
+        [switch]
+        $PickLab,
+
+        [Parameter(Mandatory = $false, HelpMessage = "Which lab properties to show a prompt for")]
+        [ValidateNotNullOrEmpty()]
+        [string[]]
+        $Properties
+    )
+
+    begin {
+
+        function LabToString($lab, $index) {
+            return "[$index]`t$($lab.Id)`t$($lab.ResourceGroupName)`t$($lab.LabName)"
+        }
+
+        $propsPassed = $PSBoundParameters.ContainsKey('Properties')
+        $pickLabPassed = $PSBoundParameters.ContainsKey('PickLab')
+
+        if($pickLabPassed) {
+           Write-Host "LABS"
+        }
+
+        $aggregateLabs = @()
+    }
+    process {
+        $aggregateLabs += $labs
+    }
+    end {
+
+        if($pickLabPassed) {
+            $index = 0
+            $aggregateLabs | ForEach-Object { Write-Host (LabToString $_ ($index++)) }
+
+            $resp = $null
+            do {
+                $resp = Read-Host -Prompt "Please select the lab to create"
+                $resp = $resp -as [int]
+                if($resp -eq $null) {
+                    Write-Host "Not an integer.Try again." -ForegroundColor red
+                }
+                if($resp -and ($resp -ge $labs.Length -or $resp -lt 0)) {
+                    Write-Host "The lab number must be between 0 and $($labs.Length - 1). Try again." -ForegroundColor red
+                    $resp = $null
+                }
+            } until ($resp -ne $null)
+            $aggregateLabs = ,$aggregateLabs[$resp]
+        }
+
+        if($propsPassed) {
+            $hash = @{}
+            $properties | ForEach-Object { $hash[$_] = Read-Host -Prompt "$_"}
+
+            $aggregateLabs = $aggregateLabs | Set-LabProperty @hash
+        }
+        return $aggregateLabs
+    }
+}
+
+# I am forced to use parameter names starting with 'An' because otherwise they get
+# bounded automatically to the fields in the CSV and added to $PSBoundParameters
+function Select-Lab {
+    [CmdletBinding()]
+    param(
+        [parameter(Mandatory = $true, HelpMessage = "Array containing one line for each lab to be created", ValueFromPipeline = $true)]
+        [ValidateNotNullOrEmpty()]
+        [psobject[]]
+        $labs,
+
+        [parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true, HelpMessage = "Id to look for")]
+        [string]
+        $AnId,
+
+        [parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true, HelpMessage = "If a lab contains any of these tags, it will be selected")]
+        [ValidateNotNullOrEmpty()]
+        [string[]]
+        $SomeTags
+    )
+
+    begin {
+        function HasAnyTags($foundTags) {
```

Comment:
It looks like "Tags" in this case isn't the Azure Resource Manager definition of tags, is that right?  Perhaps we should use maybe another term if it's a random field in the CSV that we're enabling customers to filter on.  "Category", "Label", "Info" etc...  If it's Tags like in Azure, should it be name=value pairs?  And are we expecting to write the tags to azure when creating the labs?

Reply:
Same comment as above. It is a bit weird. It is also weird to have functions specific to Tags and Id as you could craft a generic where clause in PS1 to cover queries over any field. I will leave it there for now but worth reconsidering.